junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -75,11 +88,64 @@ public void execute(ExecutionRequest request) {
 
 	private void executeAllChildren(VintageEngineDescriptor engineDescriptor,
 			EngineExecutionListener engineExecutionListener) {
+		boolean parallelExecutionEnabled = getParallelExecutionEnabled();
+
+		if (parallelExecutionEnabled) {
+			executeInParallel(engineDescriptor, engineExecutionListener);
+		}
+		else {
+			executeSequentially(engineDescriptor, engineExecutionListener);
+		}
+	}
+
+	private void executeInParallel(VintageEngineDescriptor engineDescriptor,
+			EngineExecutionListener engineExecutionListener) {
+		ExecutorService executorService = Executors.newFixedThreadPool(getThreadPoolSize());
+		RunnerExecutor runnerExecutor = new RunnerExecutor(engineExecutionListener);
+
+		List<TestDescriptor> children = new CopyOnWriteArrayList<>(engineDescriptor.getModifiableChildren());
+
+		CompletableFuture<Void> allOf = CompletableFuture.allOf(
+			children.stream().map(descriptor -> CompletableFuture.runAsync(() -> {
```

Comment:
We should also remove the descriptors from the original list to allow them to be garbage collected once they're done. I think we should be able to use an iterator on `engineDescriptor.getModifiableChildren()`, convert each into a future, call `iterator.remove` and then wait for the combined future.

Reply:
That's a good idea. I'll make the changes as you suggested!